### PR TITLE
GC.enable: raise if GC is not disabled

### LINKED ISSUE #4721: Calling `GC.enable` actually disables the garbage collector
While investigating `GC.stats` values I found that calling `GC.enable` in my test app actually disables the garbage collector, `GC.stats` values never reduce/reset and the process memory size will continually grow.

screen shot 2017-07-17 at 14 41 31

When `GC.enable` is called, `GC::stats.bytes_since_gc` will continually increase in value because the garbage collector is never run.  This is also true even if you manually call `GC.collect`.

To reproduce:
* Call `GC.enable` 1 or more times in any Crystal application
* Watch process memory size grow continually
* Watch values of `GC::stat.bytes_since_gc`

I've confirmed this on Crystal versions 0.23 && 0.23.1.

oprypin [suspects](https://gitter.im/crystal-lang/crystal?at=596cbc581c8697534a319a36):
> <oprypin> it could be crashing the GC thread or something https://github.com/ivmai/bdwgc/blob/1505f150c66ead87c275ddcef59f41d3eee80310/misc.c#L1811

The [original report](https://gitter.im/crystal-lang/crystal?at=596caf730de4d2545e36bcb9) on gitter.

### PR DESCRIPTION
**Pull request by asterite**
Fixes #4721

### CODE REVIEW

**Comment by oprypin** [comment]
Why not do the same for `disable` then?
https://github.com/ivmai/bdwgc/blob/347c1f0d0b2be5c811e7449ae2afa1b1e2d42fcc/misc.c#L1827

**Comment by asterite** [comment]
@oprypin There's no assertion there, as far as I can see. The idea is that you can invoke `disable` many times, and a counter is added, and then you need to invoke the same amount of `enable` to finally enable it. That's because you might want to do:

```crystal
GC.disable
run_something
GC.enable
```

and `run_something` might end up also disabling and enabling the GC, and without a counter that's impossible to do.

File: src/gc/boehm.cr [hunk 1]
```diff
@@ -85,6 +86,10 @@ module GC
   end
 
   def self.enable
+    unless LibGC.is_disabled != 0
+      raise "GC.enable: GC is not disabled"
```
Comment (RX14) [inline src/gc/boehm.cr]: We should already know we're in `GC.enable` from the stacktrace. We don't prefix with the method name in other parts of the stdlib at least.

Comment (asterite) [reply src/gc/boehm.cr]: The stack trace I get is:

```
GC is not disabled (Exception)
0x108ef5695: *CallStack::unwind:Array(Pointer(Void)) at ??
0x108ef5631: *CallStack#initialize:Array(Pointer(Void)) at ??
0x108ef5608: *CallStack::new:CallStack at ??
0x108ee4d55: *raise<Exception>:NoReturn at ??
0x108ee4d01: *raise<String>:NoReturn at ??
0x108f126ff: *GC::enable:Nil at ??
0x108ee47f4: __crystal_main at ??
0x108ef48c8: main at ??
```

Seems pretty cryptic to me. The call for `GC::enable` is like 6 lines below.

Comment (oprypin) [reply src/gc/boehm.cr]: It's a whole another issue, but it doesn't have to be this way.

Comment (asterite) [reply src/gc/boehm.cr]: What about "Can't enable GC: GC is not disabled". I mean, there's nothing wrong with being just a bit more verbose, right?

**Comment by lbguilherme** [comment]
I don't have an use case, but I think it is better to not raise any exception in this case and just return. `GC.enable` always finish with a enabled GC in the same way that `file.close` always finish with a closed file, even it it was already closed. This exception is not helpful as there was nothing really wrong happening.

**Comment by oprypin** [comment]
Have you read the issue? [Calling `GC.enable` actually disables the garbage collector](https://github.com/crystal-lang/crystal/issues/4721)

**Comment by lbguilherme** [comment]
Yes. I would implement it this way:

```cr
  def self.enable
    unless LibGC.is_disabled != 0
      return
    end
    LibGC.enable
  end  
```

**Comment by asterite** [comment]
But if you call `GC.enable` without calling `GC.disbale`, it's a **bug**. Bugs shouldn't be silently ignored.

**Comment by lbguilherme** [comment]
Doesn't closing a file twice fall under the same category of bug? I'm not against either behavior, just wanting to be consistent.

**Comment by asterite** [comment]
@lbguilherme 

```crystal
GC.enable
puts "Hello"
GC.disable
```

What do you expect that do to?

To me, after the last line, the GC should be enabled. But in fact, it won't be like that, with your proposal.

Closing a file twice is fine, it's an idempotent operation. Enabling a GC that's not disabled can lead to the code above, and then finding where the problem of a disabled GC is becomes super tricky.

**Comment by sdogruyol** [comment]
LGTM?

**Pull request merged**